### PR TITLE
fix(drill): specify an SA URL parm of `impersonation_target` for drill+sadrill

### PR DESCRIPTION
**Pull request by jnturton**
### SUMMARY
[Sqlalchemy-drill has been updated](https://github.com/JohnOmernik/sqlalchemy-drill/pull/74) to support impersonation with the
drill+sadrill driver, where previously it did not.  The way that callers
should specify impersonation matches that for the drill+jdbc driver in that
a SA URL parameter of impersonation_target should be set to the username
of the user to be impersonated, while the standard SA username and password
should be those of the proxy user.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
N/A

### TESTING INSTRUCTIONS
Configure a new Apache Drill connection using the drill+sadrill driver from
sqlalchemy-drill.  Enable impersonation and trigger the execution of a Drill
query.  In the Drill query profile history, confirm that the query was executed
as the impersonated user.

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by jnturton** [comment]
Bump?

File: superset/db_engine_specs/drill.py [hunk 1]
```diff
@@ -84,7 +87,7 @@ def modify_url_for_impersonation(
         if impersonate_user and username is not None:
             if url.drivername == "drill+odbc":
                 url.query["DelegationUID"] = username
-            elif url.drivername == "drill+jdbc":
+            elif url.drivername in ["drill+sadrill", "drill+jdbc"]:
                 url.query["impersonation_target"] = username
             else:
-                url.username = username
+                logger.warning("impersonation is not supported for %s", url.drivername)
```
Comment (villebro) [inline superset/db_engine_specs/drill.py:93]: IMO there's risk here that the user will *think* impersonation is enabled when using other than `odbc`, `sadrill` or `jdbc`. Would it make sense to rather raise an exception in the `else` if we're trying to impersonate and it's not supported by the driver?

File: tests/unit_tests/db_engine_specs/test_drill.py [hunk 2]
```diff
@@ -64,4 +64,4 @@ def test_sadrill_impersonation(app_context: AppContext) -> None:
     url = URL("drill+sadrill")
     username = "DoAsUser"
     DrillEngineSpec.modify_url_for_impersonation(url, True, username)
-    assert url.username == username
+    assert url.query["impersonation_target"] == username
```
Comment (villebro) [inline tests/unit_tests/db_engine_specs/test_drill.py:67]: nit: bonus points for adding impersonation tests for `drill+odbc`, `drill+jdbc` and `drill+foobar`.

**Review by villebro — COMMENTED** [review]
Thanks for this enhancement! Left a few comments, LMKWYT

File: superset/db_engine_specs/drill.py (hunk 1 above)
Comment (jnturton) [reply superset/db_engine_specs/drill.py:93]: Yes, good call.

File: tests/unit_tests/db_engine_specs/test_drill.py (hunk 2 above)
Comment (jnturton) [reply tests/unit_tests/db_engine_specs/test_drill.py:68]: The first two cases are already higher up in test_drill.py.  I've now added the drill+foobar case.

Comment (villebro) [reply tests/unit_tests/db_engine_specs/test_drill.py:68]: Apologies, I stand corrected!

File: tests/unit_tests/db_engine_specs/test_drill.py [hunk 3]
```diff
@@ -64,4 +65,27 @@ def test_sadrill_impersonation(app_context: AppContext) -> None:
...
+    url = URL("drill+foobar")
+    username = "DoAsUser"
+    exception_type = None
+
+    try:
+        DrillEngineSpec.modify_url_for_impersonation(url, True, username)
+    except Exception as e:
+        exception_type = type(e)
+
+    assert exception_type == SupersetDBAPIProgrammingError
```
Comment (villebro) [inline tests/unit_tests/db_engine_specs/test_drill.py:91]: nit: this is how it's usually done with `pytest`:
```suggestion
    with pytest.raises(SupersetDBAPIProgrammingError):
        DrillEngineSpec.modify_url_for_impersonation(url, True, username)
```

**Review by villebro — COMMENTED** [review]
LGTM, one last nit

File: tests/unit_tests/db_engine_specs/test_drill.py (hunk 3 above)
Comment (villebro) [reply tests/unit_tests/db_engine_specs/test_drill.py]: Oh - you may need to move the exception import into the test function

**Review by villebro — APPROVED** [review]
LGTM, thanks for the additional iterations!

**Pull request merged**